### PR TITLE
Add command to manually initiate SASL

### PR DESCRIPTION
**Pull request by glguy**

### CODE REVIEW

File: plugins/Admin/plugin.py [hunk 1]
```diff
@@ -359,6 +359,20 @@ def acmd(self, irc, msg, args, commandAndArgs):
...
+        @internationalizeDocstring
+        def start(self, irc, msg, args):
+            """takes no arguments
+
+            Manually initiates SASL authentication
+            """
+            if 'sasl' in irc.state.capabilities_ack:
+                irc.startSasl(msg)
+            else:
+                irc.requestCapabilities(['sasl'])
```
Comment (progval) [inline plugins/Admin/plugin.py:373]: Let's just add it to `REQUEST_CAPABILITIES` unconditionally, so you can return an error here if it's not negotiated.

File: plugins/Admin/plugin.py [hunk 2]
```diff
@@ -359,6 +359,20 @@ def acmd(self, irc, msg, args, commandAndArgs):
             self.Proxy(irc.getRealIrc(), msg, commandAndArgs)
     acmd = wrap(acmd, ['admin', many('something')])
 
+    class sasl(callbacks.Commands):
+
+        @internationalizeDocstring
+        def start(self, irc, msg, args):
```
Comment (progval) [inline plugins/Admin/plugin.py:365]: Move it to the Network plugin

File: src/irclib.py [hunk 3]
```diff
@@ -1767,6 +1757,19 @@ def resetSasl(self):
...
+        self.sasl_ecdsa_key = network_config.sasl.ecdsa_key()
+        self.sasl_scram_state = {'step': 'uninitialized'}
+        self.authenticate_decoder = None
+        self.sasl_next_mechanisms = []
+        self.sasl_current_mechanism = None
+
+        self.resetSaslMechanisms(network_config)
+
         if self.sasl_next_mechanisms:
             self.REQUEST_CAPABILITIES.add('sasl')
```
Comment (progval) [inline src/irclib.py:1774]: and you can remove that bit

File: src/irclib.py [hunk 4]
```diff
@@ -1913,6 +1916,22 @@ def _maybeStartSasl(self, msg):
                         if x.lower() in available]
             self.tryNextSaslMechanism(msg)
 
+    def startSasl(self, msg):
```
Comment (progval) [inline src/irclib.py:1919]: You should be able to dedup this function with `_maybeStartSasl` above

**Review by progval — CHANGES_REQUESTED** [review]
Thanks!

File: plugins/Admin/plugin.py (hunk 1 above)
Comment (jlu5) [reply plugins/Admin/plugin.py:373]: I would recommend renaming this to something like `sasl auth` or `sasl login`. `start` sounds like something that starts / enables a long running task, which isn't really the case here. (There's no corresponding `stop` command for it, for instance)

**Review by jlu5 — CHANGES_REQUESTED** [review]

File: plugins/Admin/plugin.py (hunk 2 above)
Comment (jlu5) [reply plugins/Admin/plugin.py:365]: I would consider renaming this to something like `sasl auth` or `sasl login`. `start` is less descriptive and sounds like something that starts a longer running task or service, which isn't the case here.

File: src/irclib.py (hunk 4 above)
Comment (glguy) [reply src/irclib.py:1901]: I *think* we should unconditionally call resetSasl() when starting a new SASL transaction as I've done here. But I'm not quite certain as this function kind of mixes configuration and the sasl state machine state initialization. Is this an issue? Do I need to split those in half?

File: plugins/Network/plugin.py [hunk 5]
```diff
@@ -306,6 +306,17 @@ def capabilities(self, irc, msg, args, otherIrc):
...
+        """takes no arguments
+
+        Manually initiate SASL authentication.
+        """
+        if 'sasl' in irc.state.capabilities_ack:
+            irc.startSasl(msg)
+            irc.replySuccess()
+        else:
+            irc.error(_('SASL not supported'))
+    saslauth = wrap(saslauth)
```
Comment (progval) [inline plugins/Network/plugin.py:319]:
```suggestion
    def authenticate(self, irc, msg, args):
        """takes no arguments

        Manually initiate SASL authentication.
        """
        if 'sasl' in irc.state.capabilities_ack:
            irc.startSasl(msg)
            irc.replySuccess()
        else:
            irc.error(_('SASL not supported'))
    authenticate = wrap(authenticate)
```

we avoid concatenating words in command names; so it has to be either a single a single word or space-separated. I don't have a strong opinion on which is best here.

**Review by progval — CHANGES_REQUESTED** [review]
One last thing

File: src/irclib.py (hunk 4 above)
Comment (progval) [reply src/irclib.py:1901]: seems fine

**Comment by progval** [comment]
Could you rewrite your history to have the test changes first, then squash the rest together? This will make history easier to read and bisect

**Comment by progval** [comment]
```diff




























diff --git a/test/test_irclib.py b/test/test_irclib.py
index 18f60b53d..2f79b2f34 100644
--- a/test/test_irclib.py
+++ b/test/test_irclib.py
@@ -1494,12 +1494,16 @@ def testBatchInterleaved(self):
 
 class SaslTestCase(SupyTestCase, CapNegMixin):
     def setUp(self):
-        pass
+        self._default_mechanisms = conf.supybot.networks.test.sasl.mechanisms()
+
+    def tearDown(self):
+        conf.supybot.networks.test.sasl.mechanisms.setValue(self._default_mechanisms)
 
     def testPlain(self):
         try:
             conf.supybot.networks.test.sasl.username.setValue('jilles')
             conf.supybot.networks.test.sasl.password.setValue('sesame')
+            conf.supybot.networks.test.sasl.mechanisms.setValue(['external', 'plain'])
             self.irc = irclib.Irc('test')
 
             self.startCapNegociation()
@@ -1526,11 +1530,11 @@ def testExternalFallbackToPlain(self):
         try:
             conf.supybot.networks.test.sasl.username.setValue('jilles')
             conf.supybot.networks.test.sasl.password.setValue('sesame')
+            conf.supybot.networks.test.sasl.mechanisms.setValue(['external', 'plain'])
             conf.supybot.networks.test.certfile.setValue('foo')
             self.irc = irclib.Irc('test')
 
             self.assertEqual(self.irc.sasl_current_mechanism, None)
-            self.irc.sasl_next_mechanisms = ['external', 'plain']
 
             self.startCapNegociation()
```

works for me after this patch

**Pull request closed without merging**